src/storage-operators/src/s3_oneshot_sink.rs

```diff
@@ -377,15 +391,18 @@ where
                                     )
                                 }
                                 row_count += u64::try_from(diff).unwrap();
-                                let uploader = s3_uploaders.entry(batch).or_insert_with(|| {
-                                    debug!(%sink_id, %worker_id, "handling batch: {}", batch);
-                                    CopyToS3Uploader::new(
-                                        sdk_config.clone(),
-                                        connection_details.clone(),
-                                        &sink_id,
-                                        batch,
-                                    )
-                                });
+                                let uploader = match s3_uploaders.entry(batch) {
+                                    Entry::Occupied(entry) => entry.into_mut(),
+                                    Entry::Vacant(entry) => {
+                                        debug!(%sink_id, %worker_id, "handling batch: {}", batch);
+                                        entry.insert(T::new(
+                                            sdk_config.clone(),
+                                            connection_details.clone(),
+                                            &sink_id,
+                                            batch,
+                                        )?)
+                                    }
+                                };
```

Comment:
can this not be done in `or_insert_with`?

Reply:
I tried that at first, but because the `new` fn is fallible I couldn't return an error from the `or_insert_with` closure...

Reply:
oh sorry, i shouldn't have missed that!